File: Solutions/98-Validate Binary Search Tree/98-Validate Binary Search Tree.js
//Third fastest
var isValidBST = function(root) {
    function validate(node, min, max) {
      if (!node) return true 
      // if (min < node.val && node.val < max) return validate(node.left, min, node.val) && validate(node.right, node.val, max)
      if (min >= node.val || node.val >= max) return false
      return validate(node.left, min, node.val) && validate(node.right, node.val, max)
    }
    return validate(root, -Infinity, Infinity)
  };

//Second fastest
function do_inorder_traversal(pNode, elements) {
    if (pNode.left) {
        do_inorder_traversal(pNode.left, elements)
    }
    elements.push(pNode.val)
    if (pNode.right) {
        do_inorder_traversal(pNode.right, elements)
    }
}

/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root) {
    let children = [];
    
    // Empty tree
    if (!root) {
        return true;
    }
    
    do_inorder_traversal(root, children);
    
    for (let i = 0; i < children.length - 1; i++) {
        if ((children[i] <= children[i-1]) || (children[i] >= children[i+1])) {
            return false;
        }
    }

    
    return true;
};

//Fastest
var isValidBST = function(root){
    if(root==null)return true;
    
    function help(root, lower, upper){
        if(root==null)return true;
        
        return (lower==null || root.val>lower)
        &&(upper==null || root.val<upper)
        && help(root.left, lower, root.val)
        && help(root.right, root.val, upper);
    }
    
    return help(root, null, null);
}
var isValidBST1 = function(root) {
    if(root==null)return true;
    
    function isGreater(root, val){
        if(root==null)return true;
        
        return root.val>val && isGreater(root.left, val)&&isGreater(root.right, val);
    }
    function isLesser(root, val){
        if(root==null)return true;
        
        return root.val<val && isLesser(root.left, val) && isLesser(root.right, val);
    }
    
    return isGreater(root.right, root.val) && isLesser(root.left, root.val) && isValidBST(root.left)&&isValidBST(root.right);
};

File: Solutions/98-Validate Binary Search Tree/98-Validate Binary Search Tree.rb
#Fourth fastest
def is_valid_bst(root, min: nil, max: nil)
    return false unless root
    val = root.val
    # p "#{val}, min: #{min}, max: #{max}"
    return false if (min && val <= min) || (max && val >= max)
    result = true
    

    if root.left
        result = root.left.val < val && is_valid_bst(root.left, max: val , min: min)
    end
    
    return false unless result
    
    if root.right
        result = root.right.val > val && is_valid_bst(root.right, min: val, max: max)
    end

    return result
end

#Third fastest
def is_valid_bst(root)
    helper(root, nil, nil)
  end
  
  def helper(root, lower, upper)
    return true unless root
    
    if lower && root.val <= lower
      return false
    end
    
    if upper && root.val >= upper
      return false
    end
    
    helper(root.left, lower, root.val) && helper(root.right, root.val, upper)
  end

#Second fastest
def is_valid_bst(root)
    return true if root.nil? 
    return false if !helper(root.left, -1.0/0.0 , root.val) or !helper(root.right, root.val, 1.0/0.0)
    return true
end

def helper(root, min, max)
    return true if root.nil?
    return false if root.val >= max or root.val <= min
    return true if helper(root.left, min, root.val) and helper(root.right, root.val, max)
    return false
end

#Fastest
$inorder_traverse = Array.new

def is_valid_bst(root)
    return true if root == nil
    $inorder_traverse.clear
	traverse_helper(root)
	sorted = true
	$inorder_traverse.each_index { |i|
		next if i == 0
		sorted = (sorted and ($inorder_traverse[i - 1] < $inorder_traverse[i]))
	}
	return sorted
end

def traverse_helper(root)
	traverse_helper(root.left) if root.left
	$inorder_traverse.push(root.val)
	traverse_helper(root.right) if root.right
end